keps/sig-network/4444-service-routing-preference/README.md

```diff
@@ -0,0 +1,998 @@
+    <th>Internal Result</th>
+  </tr>
+</thead>
+<tbody>
+  <tr>
+    <td>-</td>
+    <td>-</td>
+    <td>Auto</td>
+    <td>routingPreference=Auto</td>
+    <td>routingPreference=Auto</td>
+  </tr>
+  <tr>
+    <td>Local</td>
+    <td>-</td>
+    <td>Auto</td>
+    <td>ExternalTrafficPolicy=Local</td>
+    <td>routingPreference=Auto</td>
+  </tr>
+  <tr>
+    <td>Local</td>
+    <td>Local</td>
+    <td>Auto</td>
+    <td>ExternalTrafficPolicy=Local</td>
+    <td>InternalTrafficPolicy=Local</td>
+  </tr>
+</tbody>
+</table>
+
+### Test Plan
+
+[X] I/we understand the owners of the involved components may require updates to
+existing tests to make this code solid enough prior to committing the changes necessary
+to implement this enhancement.
+
+##### Prerequisite testing updates
+
+No updates required.
+
+##### Unit tests
+
+The main packages that will see major changes due to this enhancement are:
+- `k8s.io/kubernetes/vendor/k8s.io/endpointslice/topologycache`: `2024-01-26` -
+  `74.8`
+- `k8s.io/kubernetes/vendor/k8s.io/endpointslice`: `2024-01-26` - `80.7`
+
+There will be some code refactoring into newer packages which may prevent exact
+comparisons, but the updated packages will aim to provide equivalent coverage.
+
+The following packages will also see minor changes:
+- `k8s.io/kubernetes/pkg/controller/endpointslice`: `2024-01-26` - `61.4`
+- `k8s.io/kubernetes/pkg/proxy`: `2024-01-26` - `69.5`
+
+##### Integration tests
+
+* Verify that if both the annotation `service.kubernetes.io/topology-mode=Auto`
+  and field `routingPreference=Zone` are configured, precedence is given
+  to the annotation.
+
+##### e2e tests
+
+* Verify that EndpointSlice hints are correctly populated when
+  `routingPreference=Zone`
+* Verify through probes that when `routingPreference=Zone`, requests
+  originating from a zone which has service pods get sent to a pod in the same
+  zone. For requests originating from zones with no service pods, requests
+  should not get blackholed and should rather be forwarded to any service pod
+  from the cluster.
+
+### Graduation Criteria
+
+#### Alpha
+
+- Feature implemented behind a feature flag
+- Initial e2e tests completed and enabled
+
+### Upgrade / Downgrade Strategy
+
+No special considerations are required for upgrade / downgrade:
+
+Upon upgrade of the EndpointSlice controller (within kube-controller-manager)
+and kube-proxy:
+* If a Service had both annotation `service.kubernetes.io/topology-mode` and
+  field `routingPreference`, then the annotation will take precedence and the
+  field will be ignored. 
+* If a Service only had the annotation `service.kubernetes.io/topology-mode`,
+  then field `routingPreference`'s value will be assumed to be `Default` and the
+  behaviour will be the same as above with the annotation taking precedence.
+
+Upon downgrade of EndpointSlice controller (within kube-controller-manager) and
+kube-proxy, the `routingPreference` will simply not be considered in any
+decisions.
```

Comment:
For downgrade, how would the API server handle the `routingPreference` field that it finds on some persisted objects? Any concerns there?